daemons/attrd/attrd_cib.c

```diff
@@ -46,6 +47,81 @@ attrd_cib_destroy_cb(gpointer user_data)
     }
 }
 
+/*!
+ * \internal
+ * \brief Check a patchset change for deletion of a transient attribute
+ *
+ * \param[in] xml    Patchset change element
+ * \param[in] data  Ignored
+ *
+ * \return pcmk_rc_ok (always continue to next patchset change)
+ */
+static int
+check_deletion(xmlNode *xml, void *data)
+{
+    const char *value = NULL;
+    uint32_t node_id;
+
+    // We're only interested in deletions
+    value = crm_element_value(xml, F_CIB_OPERATION);
+    if (!pcmk__str_eq(value, "delete", pcmk__str_none)) {
+        return pcmk_rc_ok;
+    }
+    value = crm_element_value(xml, PCMK_XA_PATH);
+    if (value == NULL) {
+        crm_warn("Ignoring malformed deletion in "
+                 "CIB change notification: No " PCMK_XA_PATH);
+        return pcmk_rc_ok;
+    }
+
+    // Look for /.../node_state[ID] and get the node ID
+    value = strstr(value, "/node_state[");
+    if (value == NULL) {
+        return pcmk_rc_ok;
+    }
+    value += strlen("/node_state[");
+    if (sscanf(value, "%" SCNu32, &node_id) != 1) {
+        return pcmk_rc_ok;
+    }
+    while (isdigit(*value)) {
+        ++value;
+    }
+    if (*value != ']') {
+        return pcmk_rc_ok; // Malformed
+    }
+    ++value;
+
+    // Check whether all transient attributes were deleted
+    /* @WIP Consider whether node ID is unavailable e.g. local node could
+     * have joined after some node left the cluster and is now unknown
+     * (unfortunately the XPath doesn't have the node name)
+     */
+    if (*value == '\0') {
+        // Entire node_state was deleted
+        attrd_drop_removed_values(node_id);
+        return pcmk_rc_ok;
+    } else if (strncmp(value, "/transient_attributes",
+                       strlen("/transient_attributes")) != 0) {
+        return pcmk_rc_ok;
+    } else if (strchr(++value, '/') == NULL) {
+        // Entire transient_attributes was deleted
+        attrd_drop_removed_values(node_id);
+        return pcmk_rc_ok;
+    }
+
+    // A single transient attribute was deleted
+    value = strstr(value, "/nvpair");
+    if (value == NULL) {
+        return pcmk_rc_ok; // Malformed
+    }
+    /* @TODO The diff doesn't have the attribute name, just the nvpair
+     * id. Tracking and sync'ing nvpair IDs for attribute values should
+     * be possible. (Values removed during a rolling upgrade would
+     * never be dropped from memory, but that shouldn't be a big deal.)
+     */
```

Comment:
It's an XPath like `/cib/status/node_state[id='1']/transient_attributes[id='1']/instance_attributes[id='1']/nvpair[id='whatever']`

Reply:
Since all transient attribute requests should go through pacemaker-attrd, we should know the nvpair id, and can track it in attribute_value_t. The entire node_state or transient_attributes could be removed, but that's handled above this (we don't need to know the nvpair id).